chia/full_node/coin_store.py

```diff
@@ -179,6 +179,50 @@ async def get_coin_record(self, coin_name: bytes32) -> Optional[CoinRecord]:
                     return record
         return None
 
+    async def get_coin_records(self, names: List[bytes32]) -> List[CoinRecord]:
+        if len(names) == 0:
+            return []
+
+        coins: List[CoinRecord] = []
+        new_names: List[bytes32] = []
+        for n in names:
+            cached = self.coin_record_cache.get(n)
+            if cached is not None:
+                coins.append(cached)
+            else:
+                new_names.append(n)
+        names = new_names
+
+        if len(names) == 0:
+            return coins
+
+        async with self.db_wrapper.read_db() as conn:
+            cursors: List[Awaitable[Cursor]] = []
+            for names_chunk in chunks(names, MAX_SQLITE_PARAMETERS):
+                names_db: Tuple[Any, ...]
+                if self.db_wrapper.db_version == 2:
+                    names_db = tuple(names_chunk)
+                else:
+                    names_db = tuple([n.hex() for n in names_chunk])
+                cursors.append(
+                    conn.execute(
+                        f"SELECT confirmed_index, spent_index, coinbase, puzzle_hash, "
+                        f"coin_parent, amount, timestamp FROM coin_record "
+                        f'WHERE coin_name in ({",".join(["?"] * len(names_db))}) ',
+                        names_db,
+                    )
+                )
+
+            for cur_handle in cursors:
+                cursor = await cur_handle
```

Comment:
The intent is to do them all at a time, not concurrently right?

Reply:
I was assuming that once we issued the `execute()` call, the queries would be issued. fetching the responses is fine to do serially. it's all serially executed in the aiosqlite thread anyway.

Maybe I'm misunderstanding you. The list of coins is chunked up into the largest sqlite batch. But there may be more than one batch, so those are issued before waiting for the results.

Reply:
I think the coroutine does not start until you call await. If you want to start it sooner, you can do `asyncio.create_task(cur_handle)`.

Reply:
aiosqlite submits the queries to a thread. each `Connection` object is a thread and `execute()` submits the query to the queue of queries handled by the thread. It's a bit unclear why `execute()` itself is awaitable though. I would think the only true waiting to happen would be when asking for the results. Perhaps the queue has a size limit, in which case just submitting a job might block.

yeah, maybe it would be better to await the calls to `execute()` right away, but to collect all the cursors to let those work in parallel.

Reply:
Execute is a coroutine. I believe coroutiens are not started until you await them

Reply:
yeah, I'm a bit disappointed looking closer at this. I expected `execute()` to submit the query and return a future, but it actually submits the job and *waits* to hear back from the other thread. There doesn't seem to be a way to submit multiple jobs back-to-back without synchronizing with the thread in between (other than `executemany()`, but that can't be used for `SELECT`).